File: case2_semilla/c2_flatness_reconstruct.py
# Technique: Case 2 - Minimum Jerk QP Trajectory Planning + 6-Parameter Pseudo-Flatness Reconstruction (m11 != m22)

import numpy as np
from c2_usv_params import (
    m11_6, m22_6, m33_6, Xu_6, Yv_6, Nr_6, dP_6,
    T_MAX, T_MIN, cmd_from_thrust_array, thrust_from_cmd_poly
)
from c2_sqp_optimizer import optimize_psi_sqp

def _psi_dot_ode(psi, x_d, y_d, x_dd, y_dd, lam_tikhonov=0.015, r_hard_limit=5.0):
    u = x_d * np.cos(psi) + y_d * np.sin(psi)
    v = -x_d * np.sin(psi) + y_d * np.cos(psi)
    
    alpha = ((m22_6 - m11_6) / m22_6) * u
    beta = -x_dd * np.sin(psi) + y_dd * np.cos(psi) + (Yv_6 / m22_6) * v
    
    r = (alpha * beta) / (alpha**2 + lam_tikhonov**2)
    return float(np.clip(r, -r_hard_limit, r_hard_limit))

def reconstruct_flatness_h2(pos, vel, acc, jerk, t, psi0=None):
    x, y = pos[:, 0], pos[:, 1]
    x_d, y_d = vel[:, 0], vel[:, 1]
    x_dd, y_dd = acc[:, 0], acc[:, 1]

    # Reemplazo por la resolución algebraica (SQP)
    psi, r = optimize_psi_sqp(t, x_d, y_d, x_dd, y_dd, psi0=psi0)

    u = x_d * np.cos(psi) + y_d * np.sin(psi)
    v = -x_d * np.sin(psi) + y_d * np.cos(psi)

    u_dot = np.gradient(u, t)
    r_dot = np.gradient(r, t)

    tau_u_raw = m11_6 * u_dot - m22_6 * v * r + Xu_6 * u
    tau_r_raw = m33_6 * r_dot - (m11_6 - m22_6) * u * v + Nr_6 * r

    T1_raw = 0.5 * (tau_u_raw + tau_r_raw / dP_6)
    T2_raw = 0.5 * (tau_u_raw - tau_r_raw / dP_6)

    tau_u_max = 2.0 * T_MAX
    tau_u_min = 2.0 * T_MIN
    tau_r_max = (T_MAX - T_MIN) * dP_6
    tau_r_min = -tau_r_max

    tau_u = np.clip(tau_u_raw, tau_u_min, tau_u_max)
    tau_r = np.clip(tau_r_raw, tau_r_min, tau_r_max)

    T1_dem = np.clip(T1_raw, T_MIN, T_MAX)
    T2_dem = np.clip(T2_raw, T_MIN, T_MAX)

    cmd_1 = cmd_from_thrust_array(T1_dem)
    cmd_2 = cmd_from_thrust_array(T2_dem)

    T1_act = thrust_from_cmd_poly(cmd_1)
    T2_act = thrust_from_cmd_poly(cmd_2)

    tau_u_act = T1_act + T2_act
    tau_r_act = (T1_act - T2_act) * dP_6

    eta = np.column_stack([x, y, psi])
    nu = np.column_stack([u, v, r])
    tau_plan = np.column_stack([tau_u, tau_r])
    tau_act = np.column_stack([tau_u_act, tau_r_act])
    cmds = np.column_stack([cmd_1, cmd_2])

    return {
        'eta': eta,
        'nu': nu,
        'tau_plan': tau_plan,
        'tau_act': tau_act,
        'tau_u_raw': tau_u_raw,
        'tau_r_raw': tau_r_raw,
        'cmds': cmds,
        'T1_raw': T1_raw,
        'T2_raw': T2_raw,
        'T_plan': np.column_stack([T1_dem, T2_dem]),
        'T_act': np.column_stack([T1_act, T2_act])
    }

File: case2_semilla_c/src/c2_flatness_reconstruct.c
#include "c2_flatness_reconstruct.h"
#include "c2_sqp_optimizer.h"
#include "c2_usv_params.h"
#include <stdlib.h>
#include <math.h>

double psi_dot_ode(double psi, double x_d, double y_d, double x_dd, double y_dd) {
    const double lam_tikhonov = 0.015, r_hard_limit = 5.0;
    double u = x_d * cos(psi) + y_d * sin(psi);
    double v = -x_d * sin(psi) + y_d * cos(psi);

    double alpha = ((m22_p - m11_p) / m22_p) * u;
    double beta = -x_dd * sin(psi) + y_dd * cos(psi) + (Yv_p / m22_p) * v;

    double r = (alpha * beta) / (alpha * alpha + lam_tikhonov * lam_tikhonov);
    if (r > r_hard_limit) r = r_hard_limit;
    if (r < -r_hard_limit) r = -r_hard_limit;
    return r;
}

static void np_gradient(int N, double dt, const double *y, double *dy) {
    if (N == 1) { dy[0] = 0.0; return; }
    dy[0] = (y[1] - y[0]) / dt;
    dy[N - 1] = (y[N - 1] - y[N - 2]) / dt;
    for (int k = 1; k < N - 1; k++) dy[k] = (y[k + 1] - y[k - 1]) / (2.0 * dt);
}

void reconstruct_flatness_h2(int N, const double *t,
                              const double *x, const double *y,
                              const double *x_d, const double *y_d,
                              const double *x_dd, const double *y_dd,
                              FlatnessResult *out) {
    out->N = N;
    out->x = malloc(sizeof(double) * (size_t)N);
    out->y = malloc(sizeof(double) * (size_t)N);
    out->psi = malloc(sizeof(double) * (size_t)N);
    out->u = malloc(sizeof(double) * (size_t)N);
    out->v = malloc(sizeof(double) * (size_t)N);
    out->r = malloc(sizeof(double) * (size_t)N);
    out->tau_u = malloc(sizeof(double) * (size_t)N);
    out->tau_r = malloc(sizeof(double) * (size_t)N);
    out->tau_u_raw = malloc(sizeof(double) * (size_t)N);
    out->tau_r_raw = malloc(sizeof(double) * (size_t)N);
    out->T1_dem = malloc(sizeof(double) * (size_t)N);
    out->T2_dem = malloc(sizeof(double) * (size_t)N);
    out->cmd1 = malloc(sizeof(double) * (size_t)N);
    out->cmd2 = malloc(sizeof(double) * (size_t)N);
    out->T1_act = malloc(sizeof(double) * (size_t)N);
    out->T2_act = malloc(sizeof(double) * (size_t)N);
    out->tau_u_act = malloc(sizeof(double) * (size_t)N);
    out->tau_r_act = malloc(sizeof(double) * (size_t)N);

    for (int i = 0; i < N; i++) { out->x[i] = x[i]; out->y[i] = y[i]; }

    /* Reemplazo por la resolucion algebraica (SQP) -- no time integration. */
    optimize_psi_sqp(N, t[1] - t[0], x_d, y_d, x_dd, y_dd, out->psi, out->r);

    for (int i = 0; i < N; i++) {
        out->u[i] = x_d[i] * cos(out->psi[i]) + y_d[i] * sin(out->psi[i]);
        out->v[i] = -x_d[i] * sin(out->psi[i]) + y_d[i] * cos(out->psi[i]);
    }

    double *u_dot = malloc(sizeof(double) * (size_t)N);
    double *r_dot = malloc(sizeof(double) * (size_t)N);
    double dt = t[1] - t[0];
    np_gradient(N, dt, out->u, u_dot);
    np_gradient(N, dt, out->r, r_dot);

    double tau_u_max = 2.0 * T_MAX, tau_u_min = 2.0 * T_MIN;
    double tau_r_max = (T_MAX - T_MIN) * dP_p, tau_r_min = -tau_r_max;

    for (int i = 0; i < N; i++) {
        double tau_u_raw = m11_p * u_dot[i] - m22_p * out->v[i] * out->r[i] + Xu_p * out->u[i];
        double tau_r_raw = m33_p * r_dot[i] - (m11_p - m22_p) * out->u[i] * out->v[i] + Nr_p * out->r[i];
        out->tau_u_raw[i] = tau_u_raw;
        out->tau_r_raw[i] = tau_r_raw;

        double tu = tau_u_raw, tr = tau_r_raw;
        if (tu > tau_u_max) tu = tau_u_max;
        if (tu < tau_u_min) tu = tau_u_min;
        if (tr > tau_r_max) tr = tau_r_max;
        if (tr < tau_r_min) tr = tau_r_min;
        out->tau_u[i] = tu;
        out->tau_r[i] = tr;

        double T1_raw = 0.5 * (tau_u_raw + tau_r_raw / dP_p);
        double T2_raw = 0.5 * (tau_u_raw - tau_r_raw / dP_p);
        double T1d = T1_raw, T2d = T2_raw;
        if (T1d > T_MAX) T1d = T_MAX;
        if (T1d < T_MIN) T1d = T_MIN;
        if (T2d > T_MAX) T2d = T_MAX;
        if (T2d < T_MIN) T2d = T_MIN;
        out->T1_dem[i] = T1d;
        out->T2_dem[i] = T2d;

        double cmd1 = cmd_from_thrust_richards(T1d);
        double cmd2 = cmd_from_thrust_richards(T2d);
        out->cmd1[i] = cmd1;
        out->cmd2[i] = cmd2;

        double T1_act = thrust_from_cmd_richards(cmd1);
        double T2_act = thrust_from_cmd_richards(cmd2);
        out->T1_act[i] = T1_act;
        out->T2_act[i] = T2_act;

        out->tau_u_act[i] = T1_act + T2_act;
        out->tau_r_act[i] = (T1_act - T2_act) * dP_p;
    }

    free(u_dot); free(r_dot);
}

void flatness_result_free(FlatnessResult *r) {
    free(r->x); free(r->y); free(r->psi);
    free(r->u); free(r->v); free(r->r);
    free(r->tau_u); free(r->tau_r);
    free(r->tau_u_raw); free(r->tau_r_raw);
    free(r->T1_dem); free(r->T2_dem);
    free(r->cmd1); free(r->cmd2);
    free(r->T1_act); free(r->T2_act);
    free(r->tau_u_act); free(r->tau_r_act);
}
